### PR TITLE
Update haproxy and keepalived config

### LINKED ISSUE #3052: High Availability Considerations docs outdated
## Is this a BUG REPORT or FEATURE REQUEST?

Choose one: BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`):
```
kubeadm version: &version.Info{Major:"1", Minor:"29", GitVersion:"v1.29.3", GitCommit:"6813625b7cd706db5bc7388921be03071e1a492d", GitTreeState:"clean", BuildDate:"2024-03-15T00:06:16Z", GoVersion:"go1.21.8", Compiler:"gc", Platform:"linux/amd64"}
```
**Environment**:
- **Kubernetes version** (use `kubectl version`):
```
Client Version: v1.29.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.2
```
- **Cloud provider or hardware configuration**:
Ubuntu VMs
- **OS** (e.g. from /etc/os-release):
```
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```
- **Kernel** (e.g. `uname -a`):
```
Linux vmubtkube-a01 5.15.0-1054-kvm #59-Ubuntu SMP Thu Mar 14 16:03:41 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```
- **Container runtime (CRI)** (e.g. containerd, cri-o):
cri-o
- **Container networking plugin (CNI)** (e.g. Calico, Cilium):
Calico
- **Others**:

## What happened?
haproxy was still including nodes which returned non 200 health checks. I attempted to troubleshoot but there's been significant changes since haproxy v2.1 so documentation isn't readily available. It seems most likely the health check was running over HTTP (plaintext) and ignoring the returned 400 error code. In addition, haproxy v2.1 is no longer officially supported.

I also observed the very low timeouts in haproxy lead to frequent termination of `kubectl ... -w` and `kubectl logs -f`

Edit: I think it may also be possible that `ssl-hello-chk` option in the guide is overriding `httpchk` which would also explain the behavior I was seeing. https://cbonte.github.io/haproxy-dconv/2.1/configuration.html#5.2-check

https://endoflife.date/haproxy

## What you expected to happen?
The guide to provide a setup using software still supported (patched) by the vendor

## How to reproduce it (as minimally and precisely as possible)?
Attempt to use config in haproxy v2.8 docs (currently the default LTS) and it fails due to syntax changes

## Anything else we need to know?
For keepalived check, I ended up with
```
#!/bin/sh

errorExit() {
    echo "*** $*" 1>&2
    exit 1
}

curl -sfk --max-time 2 https://localhost:5443/healthz -o /dev/null   || errorExit "Error GET https://localhost:5443/healthz
```

It's unclear in the guide why haproxy is only being health checked if it's running on the VIP. The guide configuration could allow keepalived to move the VIP to a node with working API server and broken haproxy which will fail and it will be moved again. Additionally, it doesn't check the health endpoint so the VIP could be moved to a node that's misconfigured but still responds to requests.

For haproxy 2.8, I ended up with the following (static pod)
```
global
    log stdout format raw local0
    daemon

#---------------------------------------------------------------------
# common defaults that all the 'listen' and 'backend' sections will
# use if not designated in their block
#---------------------------------------------------------------------
defaults
    mode                    http
    log                     global
    option                  httplog
    option                  dontlognull
    option http-server-close
    option forwardfor       except 127.0.0.0/8
    option                  redispatch
    retries                 1
    timeout http-request    10s
    timeout queue           20s
    timeout connect         5s
    timeout client          1800s
    timeout server          1800s
    timeout http-keep-alive 10s
    timeout check           10s

#---------------------------------------------------------------------
# apiserver frontend which proxys to the control plane nodes
#---------------------------------------------------------------------
frontend apiserver
    bind *:5443
    mode tcp
    option tcplog
    default_backend apiserverbackend

#---------------------------------------------------------------------
# round robin balancing for apiserver
#---------------------------------------------------------------------
backend apiserverbackend
    option httpchk

    http-check connect ssl
    http-check send meth GET uri /healthz
    http-check expect status 200

    mode tcp
    balance roundrobin
    server vmubtkube-a01 172.16.1.65:6443 check verify none
    server vmubtkube-a02 172.16.1.66:6443 check verify none
    server vmubtkube-a03 172.16.1.67:6443 check verify none
```

I don't know much about haproxy, but I'm not sure how the health check worked before unless haproxy v2.1 didn't validate certificates by default.

I reached these conclusions with the following tests:
- kill etcd on a control plane node rendering api-server inoperable (it responds 500 to /healthz)
- kill haproxy on a node while the control plane is running (other haproxy instances should still route to the node since api-server is operable)
- reboot the node with the VIP, assuming it takes a non trivial amount of time (30 seconds or more), VIP should move to a new node and haproxy running on existing nodes should remove the rebooting node api-server. On completion of reboot, haproxy instances should be available on all nodes with all api-server backends active

### PR DESCRIPTION
**Pull request by nijave**
Closes #3052

### CODE REVIEW

File: docs/ha-considerations.md [hunk 1]
```diff
@@ -131,8 +125,8 @@ defaults
     timeout http-request    10s
     timeout queue           20s
     timeout connect         5s
-    timeout client          20s
-    timeout server          20s
+    timeout client          1800s
+    timeout server          1800s
```
Comment (neolit123) [inline docs/ha-considerations.md:129]: what is the significance of increasing these timeouts? can we get away with the shorter ones?

File: docs/ha-considerations.md [hunk 2]
```diff
@@ -91,28 +91,22 @@ errorExit() {
...
-if ip addr | grep -q ${APISERVER_VIP}; then
-    curl --silent --max-time 2 --insecure https://${APISERVER_VIP}:${APISERVER_DEST_PORT}/ -o /dev/null || errorExit "Error GET https://${APISERVER_VIP}:${APISERVER_DEST_PORT}/"
-fi
+curl -sfk --max-time 2 https://localhost:${APISERVER_DEST_PORT}/healthz -o /dev/null || errorExit "Error GET https://localhost:${APISERVER_DEST_PORT}/healthz"
 ```
 
-There are some placeholders in `bash` variable style to fill in:
-- `${APISERVER_VIP}` is the virtual IP address negotiated between the `keepalived` cluster hosts.
-- `${APISERVER_DEST_PORT}` the port through which Kubernetes will talk to the API Server.
+Fill in the placeholder `${APISERVER_DEST_PORT}` with the port through which Kubernetes will talk to the API Server (the port haproxy or your load balancer will be listening on).
```
Comment (neolit123) [inline docs/ha-considerations.md:97]:
```suggestion
Fill in the placeholder `${APISERVER_DEST_PORT}` with the port through which Kubernetes will talk to the API Server. That is the port haproxy or your load balancer will be listening on.
```

**Review by neolit123 — APPROVED** [review]
thanks
/approve
/kind documentation

/cc @mbert

**Review by mbert — COMMENTED** [review]
Looks good to me. As the timeout values are concerned, the reason for increasing them was given in the original ticket, however I'd still be interested to understand why they need to be that high.

**Comment by nijave** [comment]
> Looks good to me. As the timeout values are concerned, the reason for increasing them was given in the original ticket, however I'd still be interested to understand why they need to be that high.

I'm not sure they necessarily need to be that high--I'm not sure if Kubernetes does any sort of keepalives but they would ideally be set around what k8s has

**Comment by nijave** [comment]
I found https://github.com/kubernetes/kubernetes/pull/95981 which would explain the 30s interval I'm seeing. I also asked in Kubernetes Slack https://kubernetes.slack.com/archives/C09NXKJKA/p1714925405232349

With that info, maybe 35s is more reasonable?

**Comment by mbert** [comment]
> I found [kubernetes/kubernetes#95981](https://github.com/kubernetes/kubernetes/pull/95981) which would explain the 30s interval I'm seeing. I also asked in Kubernetes Slack https://kubernetes.slack.com/archives/C09NXKJKA/p1714925405232349
> 
> With that info, maybe 35s is more reasonable?

Yes, definitely.

**Review by neolit123 — COMMENTED** [review]
please squash the commits to one, then this can merge.

**Review by neolit123 — COMMENTED** [review]
/lgtm
thanks for the changes!

**Comment by neolit123** [comment]
also thanks for the discussion about this @mbert @nijave

**Pull request merged**